docs/spec/overload.rst

```diff
@@ -99,8 +109,376 @@ variable is not sufficient.
+
+      # Step 1 eliminates first overload. Step 4 and
+      # step 5 do not apply. Step 6 picks the second
+      # overload.
+      r2 = example3(1, 2)
+      reveal_type(r2)  # Should reveal tuple[int, int]
+
+      # Step 1 doesn't eliminate any overloads. Step 4
+      # picks the third overload.
+      val = [1, 2, 3, 4]
+      r3 = example3(*val)
+      reveal_type(r3)  # Should reveal tuple[int, ...]
+
+
+Example 4::
+
+  @overload
+  def example4(x: list[int], y: int) -> int: ...
+  @overload
+  def example4(x: list[str], y: str) -> int: ...
+  @overload
+  def example4(x: int, y: int) -> list[int]: ...
+
+  def test(v1: list[Any], v2: Any):
+      # Step 4 eliminates third overload. Step 5
+      # determines that first and second overloads
+      # both apply and are ambiguous due to Any, but
+      # return types are consistent.
+      r1 = example4(v1, 1)
+      reveal_type(r1)  # Should reveal int
+
+      # Step 4 eliminates third overload. Step 5
+      # determines that first and second overloads
+      # both apply and are ambiguous due to Any, but
+      # return types are consistent.
+      r2 = example4([1], v2)
+      reveal_type(r2)  # Should reveal int
+
+      # Step 5 determines that first, second, and third
+      # overloads all apply and are ambiguous due to Any.
+      # These have inconsistent return types, so evaluated
+      # type is Any.
+      r3 = example4(v2, v2)
+      reveal_type(r3)  # Should reveal Any
+
+
+
+Argument type expansion
+^^^^^^^^^^^^^^^^^^^^^^^
+
+When performing argument type expansion, the following types should be
+expanded:
+
+1. Unions: Each subtype of the union should be considered as a separate
+argument type. For example, the type ``int | str`` should be expanded
+into ``int`` and ``str``.
+
+2. ``bool`` should be expanded into ``Literal[True]`` and ``Literal[False]``.
+
+3. ``Enum`` types (other than those that derive from ``enum.Flag``) should
+be expanded into their literal members.
+
+4. ``type[A | B]`` should be expanded into ``type[A]`` and ``type[B]``.
+
+5. Tuples of known length that contain expandable types should be expanded
+into all possible combinations of their subtypes. For example, the type
+``tuple[int | str, bool]`` should be expanded into ``(int, Literal[True])``,
+``(int, Literal[False])``, ``(str, Literal[True])``, and
+``(str, Literal[False])``.
+
+
+[Eric's note for reviewers: I'm not 100% convinced we should
+support argument expansion in all of these cases. Tuple expansion,
+in particular, can be very costly and can quickly blow up in complexity.
+Currently, pyright and mypy support only the case 1 in the list above,
+but I have had requests to support 2 and 3.]
+
+When performing type expansion for an argument, the argument that
+is targeted for expansion should be evaluated without the use of
+any context. All arguments that are not yet expanded should
+continue to be evaluated with the benefit of context supplied by parameter
```

Comment:
This (the use of context in evaluation) seems important and non-obvious (at least to me). I'd like to see it covered in its own section rather than only mentioned as an aside in argument type expansion.

Reply:
Yeah, I agree that it's important for consistent overload call evaluation. It should eventually be covered in more detail in other parts of the spec, but I'd prefer not to tackle that as part of the overload updates. This is already going to be a difficult set of updates to land. I'm not sure how to separate these other than to allude to context-based evaluation for now.